docs/content/terraform.md

```diff
@@ -681,3 +682,194 @@ Keep in mind that it's up to you to decide how to use OPA's Terraform tests and
 * Embed it into your deployment system to catch problems that arise when applying Terraform to production after applying it to staging
+
+If you'd like to explore an additional example that uses terraform modules please continue below.
+
+# Working with Modules
+
+## Module Steps
+
+### 1. Create and save Terraform module plan 
+
+Create a new Terraform file that includes a
+security group and security group from a module.  (This example uses the module from https://github.com/terraform-aws-modules/terraform-aws-security-group)
+
+```shell
+cat >main.tf <<EOF
+provider "aws" {
+  region = "us-east-1"
+}
+
+data "aws_vpc" "default" {
+  default = true
+}
+
+module "http_sg" {
+  source = "git::https://github.com/terraform-aws-modules/terraform-aws-security-group.git?ref=v3.10.0"
+
+  name        = "http-sg"
+  description = "Security group with HTTP ports open for everybody (IPv4 CIDR), egress ports are all world open"
+  vpc_id      = data.aws_vpc.default.id
+
+  ingress_cidr_blocks = ["0.0.0.0/0"]
+}
+
+
+resource "aws_security_group" "allow_tls" {
+  name        = "allow_tls"
+  description = "Allow TLS inbound traffic"
+  vpc_id      = data.aws_vpc.default.id
+
+  ingress {
+    description = "TLS from VPC"
+    from_port   = 443
+    to_port     = 443
+    protocol    = "tcp"
+    cidr_blocks = ["10.0.0.0/8"]
+  }
+
+  egress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+
+  tags = {
+    Name = "allow_tls"
+  }
+}
+EOF
+```
+
+Then initialize Terraform and ask it to calculate what changes it will make and store the output in `tfplan.binary`.
+
+```shell
+terraform init
+terraform plan --out tfplan.binary
+```
+
+### 2. Convert the new Terraform plan into JSON
+
+Use the Terraform show command to produce the json representation of the terraform plan
+
+```shell
+terraform show -json tfplan.binary > tfplan2.json
+```
+
+### 3. Write the OPA policy to collect resources
+
+The policy evaluates if a security group is valid based on the contents of it's description:
+
+* Resources can be specified under the root module or in child modules
+* We want to evaluate against the combined group of these resources
+* This example is scoped to the planned changes section of the json representation
+
+The policy uses the walk keyword to explore the json structure, and uses conditions to filter for the specific paths where resources would be found.
+
+**terraform_module.rego**:
+
+```shell
+package terraform.module
+
+deny[msg] {
+  desc := resources[r].values.description
+  contains(desc, "HTTP")
+  msg = sprintf("No security groups should be using HTTP. Resource in violation: %v",[r.address])
+}
+
+resources := { r |
+  some path, value
+    
+  # Walk over the JSON tree and check if the node we are
+  # currently on is a module (either root or child) resources
+  # value.
+  walk(input,planned_changes, [path, value])
```

Comment:
Running through the example quickly before merging I ran into an eval error, looks like this should be `input.planned_values`?

Reply:
yep bad copy and paste, on my end, pushing that fix up now

Reply:
fixed!